javatests/arcs/core/analysis/RecipeGraphFixpointIteratorTest.kt

```diff
@@ -0,0 +1,267 @@
+    } ?: input
+
+    override fun transform(
+        particle: Recipe.Particle,
+        handle: Recipe.Handle,
+        spec: HandleConnectionSpec,
+        input: AbstractSet<String>
+    ) = input.set?.let {
+        AbstractSet<String>(it + "p:${particle.spec.name} -> h:${handle.name}")
+    } ?: input
+}
+
+@RunWith(JUnit4::class)
+class RecipeGraphFixpointIteratorTest {
+    private val thing = Recipe.Handle("thing", Recipe.Handle.Fate.CREATE, TypeVariable("thing"))
+    private val name = Recipe.Handle("name", Recipe.Handle.Fate.CREATE, TypeVariable("name"))
+    private val readCnxn = HandleConnectionSpec("r", HandleMode.Read, TypeVariable("r"))
+    private val writeCnxn = HandleConnectionSpec("w", HandleMode.Write, TypeVariable("w"))
+    private val readerSpec = ParticleSpec(
+        "Reader",
+        listOf(readCnxn).associateBy { it.name },
+        "ReaderLocation"
+    )
+    private val writerSpec = ParticleSpec(
+        "Writer",
+        listOf(writeCnxn).associateBy { it.name },
+        "WriterLocation"
+    )
+    private val anotherWriterSpec = ParticleSpec(
+        "AnotherWriter",
+        listOf(writeCnxn).associateBy { it.name },
+        "WriterLocation"
+    )
+    private val readerParticle = Recipe.Particle(
+        readerSpec,
+        listOf(Recipe.Particle.HandleConnection(readCnxn, thing))
+    )
+    private val writerParticle = Recipe.Particle(
+        writerSpec,
+        listOf(Recipe.Particle.HandleConnection(writeCnxn, thing))
+    )
+    private val anotherWriterParticle = Recipe.Particle(
+        anotherWriterSpec,
+        listOf(Recipe.Particle.HandleConnection(writeCnxn, thing))
+    )
+
+    @Test
+    fun straightLineFlow() {
+        // [Writer] -> (thing) -> [Reader]
+        val recipe = Recipe(
+            "StraightLine",
+            listOf(thing).associateBy { it.name },
+            listOf(readerParticle, writerParticle),
+            "arcId"
+        )
+        val graph = RecipeGraph(recipe)
+        val analyzer = TestAnalyzer(setOf("p:${writerParticle.spec.name}"))
+        analyzer.computeFixpoint(graph)
+        assertThat(analyzer.getValue(writerParticle)?.set).isEmpty()
+        assertThat(analyzer.getValue(thing)?.set)
+            .containsExactly("p:Writer", "p:Writer -> h:thing")
+        assertThat(analyzer.getValue(readerParticle)?.set)
+            .containsExactly("p:Writer", "p:Writer -> h:thing", "h:thing", "h:thing -> p:Reader")
+    }
+
+    @Test
+    fun joinsFlow() {
+        // [Writer] ---------> (thing) -> [Reader]
+        // [AnotherWriter] ------^
+        val recipe = Recipe(
+            "Join",
+            listOf(thing).associateBy { it.name },
+            listOf(readerParticle, writerParticle, anotherWriterParticle),
+            "arcId"
+        )
+        val graph = RecipeGraph(recipe)
+        val analyzer = TestAnalyzer(
+            setOf("p:${writerParticle.spec.name}", "p:${anotherWriterParticle.spec.name}")
+        )
```

Comment:
You might want to consider writing a helper method for parts of this setup code. It's pretty boilerplatey, and duplicated a lot, and not essential to understanding the test. 

e.g. This is a bit nicer:
```
val graph = createGraph(
    name = "Join",
    handles = listOf(thing),
    particles = listOf(...)),
)
```

Reply:
Like the idea. Done.